### PR TITLE
RPM build and Makefile targets for sources

### PR DESCRIPTION
**Pull request by Rudd-O**
Supersedes https://github.com/goatpig/BitcoinArmory/pull/55.

### CODE REVIEW

**Comment by goatpig** [comment]
I'm uncomfortable with this PR, can you walk me through the changes?

**Comment by Rudd-O** [comment]
Yes.

The first change factors out the version utilities to a separate file so that they may be reused without having to import ArmoryUtils.

The second change adds the specfile and the Makefile targets needed to do.

The third change spins off the program that creates the tarball RPM wants to a separate command line utility.

That's really all there is, nothing more.

**Comment by Rudd-O** [comment]
I have rebased on top of your dev.

**Comment by Rudd-O** [comment]
Dude, why hasn't this been merged?

**Comment by goatpig** [comment]
I plan to overhaul the build process on *nix with autoconf for the next release. I'll reassess this change then.

**Comment by Rudd-O** [comment]
Brother, it's been about a year and two months, and you didn't merge this.  So many people using Red Hat or Fedora could have built packages for their systems using this PR during that time.  It made no sense to wait until autoconf is in place — the way it was, this patch worked fine, and it could have been updated to autoconf with a two lines' change.

I'm sad about this.

**Comment by Rudd-O** [comment]
Actually, it's far worse than I thought.  Since the move to autoconf / automake, it's not possible to do a `make` anymore, because it fails with the fcgi submodule not being cloned.  Worse still, even when the fcgi submodule is cloned, `make distclean` doesn't even work.  What's the point of using autotools if half the targets don't work?

Totally the wrong direction for the project.  I'm considering simply abandoning this wallet.

**Comment by droark** [comment]
@Rudd-O - What's the alternative to all of this? The original build system was held together by chewing gum and was part of the technical debt that ATI was going to clean up before things went south. goatpig did the best he could, seeing as how nobody was exactly offering comprehensive patches to get us onto whatever build system we're supposed to be on in 2018. automake/autoconf, while definitely not ideal, is something we're relatively comfortable handling, even if there is an element of "tinker 'til it works" that really shouldn't be there. In fact, I had an intern who designed an automake/autoconf system that was going to be part of the unreleased/ATI version of 0.94. We can't use it for this project but I remembered enough from the project that I was able to work with goatpig to get the build system basically up and running.

Also, if this could've been updated for autoconf with two lines, why not say something when goatpig brought up his reasoning? For whatever reasons (time, I assume), goatpig tends to require a bit of pushing in order for PRs to be accepted. Even I have to push a bit, and I spent two years working with the guy! If it had been made clear that this patch wouldn't have KOed the new build system, perhaps it would've been accepted one way or another. I get that this is frustrating on some level. I'm just saying that there wasn't a serious push to get the PR merged.

FWIW, I'm willing to look into getting this across the finish line. That said, I don't use RH/Fedora/CentOS, so I have to have somebody make sure any changes I make don't cause problems elsewhere.

**Comment by droark** [comment]
@Rudd-O - I'd like to check one last time and see if you're willing to work with me on getting this PR pushed through. I'm cleaning up the Autotools files a bit and think most of your changes look sane. If you're willing to actually use the code in this PR, I *will* help. Maybe not overnight but I will do what I can to help guide this through.

(That said, it's Autotools or bust. It has taken me quite awhile to become comfortable with Autotools. I'm not learning yet another build system, at least not yet! :) )

**Comment by Rudd-O** [comment]
What do you need from me?  I'm happy to help.

**Comment by droark** [comment]
Okay. I'll cover most of this in code comments. For now, this needs to be rebased. Note that cppForSwig/Makefile.am may be shifting somewhat as I clean it up. I'll help guide you through any relevant changes.

Also, this PR needs to be tested against Ubuntu if it hasn't been. Honestly, the only Linux distro we fully support is Ubuntu. If we can support other distros without significant effort, awesome, but Ubuntu is the baseline. Any solution that messes with Ubuntu won't fly.

Finally, just FYI, if this gets in, it'll probably be in 0.97 at the earliest. Plates are full all around.

Thanks.

File: BitcoinArmory.spec.in [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# disable debug package, otherwise nasty build error in fedora >= 23
```
Comment (droark) [inline BitcoinArmory.spec.in]: As mentioned elsewhere, I'm not familiar with Fedora's build system. Do the spec files help compile Fedora packages?

File: BitcoinArmory.spec.in [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+Name:     BitcoinArmory
+Version:  @PACKAGE_VERSION@
+Release:  %{mybuildnumber}%{?dist}
+Summary:  A high-security Bitcoin wallet
+Group:    Applications/Financial
+License:  MIT
+Source0:  bitcoinarmory-%{version}.tar.gz
+
+BuildRequires: python2-twisted, python2-psutil, swig, python2-devel, gcc-c++, PyQt4-devel, rsync
```
Comment (droark) [inline BitcoinArmory.spec.in]: Requires may be changing soon. Just FYI.

File: Makefile.am [hunk 3]
```diff
@@ -5,7 +5,7 @@ EXTRA_DIST = *.py *.md LICENSE LICENSE-ATI LICENSE-MIT \
 	nginx_example.conf imgList.xml autogen.sh \
 	armoryengine/*.py ui/*.py \
 	img/* lang/* dpkgfiles/* \
-	windowsbuild/* linuxbuild/* osxbuild/*
+	windowsbuild/* linuxbuild/* osxbuild/* BitcoinArmory.spec
```
Comment (droark) [inline Makefile.am]: Does the spec need to be distributed as a standalone file when packaged under Fedora?

File: Makefile.am [hunk 4]
```diff
@@ -66,10 +66,16 @@ endif
 
 # Skip Linux-specific steps on OSX.
 if ! BUILD_DARWIN
+# Unfuck broken build system that, upon make install, incorrectly
+# regenerates files with DESTDIR instead of prefix.
+# Who the hell uses autotools in 2017?
```
Comment (droark) [inline Makefile.am]: We do. :P Cut this line.

File: Makefile.am [hunk 5]
```diff
@@ -66,10 +66,16 @@ endif
...
 if HAVE_GUI
 	rsync -rupE --exclude="img/.DS_Store" img $(DESTDIR)$(prefix)/share/armory/
 	sed "s: /usr: $(prefix):g" < dpkgfiles/armory > $(DESTDIR)$(prefix)/bin/armory
 	chmod +x $(DESTDIR)$(prefix)/bin/armory
+	mkdir -p $(DESTDIR)$(datadir)/applications
```
Comment (droark) [inline Makefile.am]: What is being placed under this subdirectory?

File: configure.ac [hunk 6]
```diff
@@ -164,7 +164,8 @@ AM_CONDITIONAL([BUILD_DARWIN], [test x$BUILD_OS = xdarwin])
 
 AC_CONFIG_FILES(Makefile
 		cppForSwig/Makefile
-		cppForSwig/lmdb/Makefile)
+		cppForSwig/lmdb/Makefile
+		BitcoinArmory.spec)
```
Comment (droark) [inline configure.ac]: This looks like it'll interfere with Ubuntu. Am I right?

**Pull request closed without merging**